site/_posts/2020-07-23-securing-react-native-with-oauth.md

```diff
@@ -0,0 +1,480 @@
+
+```javascript
+//...
+const handleAuthorize = useCallback(
+  async provider => {
+    try {
+      const newAuthState = await authorize(configs.fusionauth);
+
+      setAuthState({
+        hasLoggedInOnce: true,
+        ...newAuthState
+      });
+    } catch (error) {
+      Alert.alert('Failed to log in', error.message);
+    }
+  },
+  [authState]
+);
+//...
+```
+
+## Building the user interface
+
+Next, we need to create a user interface (UI). We'll create a simple UI to display the access token and any other server response data. The access token is what FusionAuth provides once a user has authenticated. The code for UI implementation looks like this:
+
+```react
+//...
+return (
+  <View style={styles.container}>
+    <Image
+      source={require('./fusionauth.png')}
+    />
+    {authState.accessToken ? (
+      <View style={styles.userInfo}>
+        <View>
+          <Text style={styles.userInfoText}>
+            accessToken
+          </Text>
+          <Text style={styles.userInfoText}>
+            {authState.accessToken}
+          </Text>
+          <Text style={styles.userInfoText}>
+            accessTokenExpirationDate
+          </Text>
+          <Text style={styles.userInfoText}>
+            {authState.accessTokenExpirationDate}
+          </Text>
+        </View>
+      </View>
+    ) : (
+        <TouchableOpacity
+          style={styles.button}
+          onPress={() => handleAuthorize()}
+        >
+          <Text style={styles.buttonText}>
+            Login with FusionAuth</Text>
+          </TouchableOpacity>
+      )}
+  </View>
+);
+```
+
+We'll show one of two states, depending on whether we have an `accessToken`. At this point, we can run the app in the simulator:
+
+<div class="d-flex justify-content-center mb-5 mt-1 youtube">
+<iframe width="560" height="315" src="https://www.youtube.com/embed/rmrqXT30X38" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+</div>
+
+Following best practices, the mobile application opens up a system browser for user authentication, rather than a webview or embedded user-agent.
+
+## Securely storing the token
+
+Once the user has successfully authenticated, we will have a JWT, and possibly a refresh token, which should be stored securely. Storing sensitive data like an access token in `Asyncstorage` is bad practice. We can use another third-party package to access the iOS Keychain and Android secure storage. 
+
+There are many options, but the Formidable team, the creators of the `react-native-app-auth` package we are using, recommend [`react-native-keychain`](https://github.com/oblador/react-native-keychain). Install it by running the following command:
+
+```shell
+yarn add react-native-keychain
+```
+
+We need to store the access token after successful authentication. We can do so by using this code:
+
+```javascript
+//...
```

Comment:
Where does this code go to make it work? I'm a little lost on how this plugs in.